nodejs-env/src/main/scala/org/scalajs/jsenv/nodejs/NodeJSEnv.scala

```diff
@@ -164,6 +165,37 @@ object NodeJSEnv {
         case Input.CommonJSModulesToLoad(modules) =>
           for (module <- modules)
             writeRequire(module)
+
+        case Input.ESModulesToLoad(modules) =>
+          if (modules.nonEmpty) {
+            val uris = modules.map {
+              case module: FileVirtualBinaryFile =>
+                module.file.toURI
+              case module =>
+                tmpFile(module.path, module.inputStream).toURI
+            }
```

Comment:
Isn't there a requirement on the filename? If yes, you should check it.

Reply:
[Not really](https://nodejs.org/api/esm.html#esm_interop_with_existing_modules), although if it doesn't have the extension `.mjs` it will be parsed as a CommonJS module, then *exposed* as an ES module. So it can do something meaningful, and we should not prevent users from using that behavior of Node.js, IMO, at least as long as it's experimental in Node.js: giving more freedom to user means they are less dependent on us releasing new versions if and when Node.js changes. I suggest we reconsider once Node.js settles on a final design.

Reply:
To give a bit more sense to the behavior, I was also considering renaming the 3 `Input` types with a more "action"-oriented name, such as:

* `RunScripts`
* `ImportModules`
* `RequireModules`

Then it's clearer what those operations map to, and that the burden of making sure the target files make sense given the JSEnv's peculiarities is in the hands of the user.

Reply:
I think we are thinking about this the wrong way around. We should specify what the file contains, not how to load it. Otherwise the `Input`'s semantics is not independent of the `JSEnv` anymore.

So I would propose the opposite: Irrespective of the extension of the virtual file, interpret is as specified in the input, renaming it if necessary.

Further, it seems that we will always have a need to load multiple things, so maybe we should change `Input` to a single file but repeated. (Although this will make it harder to add an "environment" with other scripts that are available for loading).

In any case, I'm OK with `ESModulesToLoad` in this PR, but I would like to see some effort to avoid proliferation of different `Input` types going forward; as it stands now, I think we are adding them too easily given the burden they put on downstream maintainers (including us).

Lastly, I just realized we should make `NodeJSEnv` more aware of the coupling of `--experimental-modules` and `ESModulesToLoad`. Maybe enable it automatically or give it a higher level option, failing if it is not set.

Reply:
Renaming is not possible. That would break links from other files to the renamed file, which can go up to loading twice the same module (because it's *not* the same module anymore if it doesn't have the same URL).

> Further, it seems that we will always have a need to load multiple things, so maybe we should change Input to a single file but repeated.

Probably, yes.

> (Although this will make it harder to add an "environment" with other scripts that are available for loading).

I don't think we need that anyway.

> In any case, I'm OK with ESModulesToLoad in this PR, but I would like to see some effort to avoid proliferation of different Input types going forward; as it stands now, I think we are adding them too easily given the burden they put on downstream maintainers (including us).

I've filed #3506 so that we revisit that.

> Lastly, I just realized we should make NodeJSEnv more aware of the coupling of `--experimental-modules` and `ESModulesToLoad`. Maybe enable it automatically or give it a higher level option, failing if it is not set.

I really don't think so. If say Node.js 12 suddenly supports modules by default, and drops the flag (which they *do* tend to do), we'll block users from upgrading to Node.js 12 until we can release a new version of Scala.js that does not add/require `--experimental-modules` (and if we did add it automatically before, that would break earlier versions of Node.js). Again, as long as it's experimental in Node.js, I would leave `NodeJSEnv` as dumb as possible and let users fill in the blanks.

Reply:
> Renaming is not possible. That would break links from other files to the renamed file, which can go up to loading twice the same module (because it's not the same module anymore if it doesn't have the same URL).

Hmmm... that is fair. Maybe it is entirely wrong to allow virtual files in there. Maybe we should take file names instead and allow a different input type that is literal code...

Let's go ahead with this and defer to #3506 then.